eth-contracts/contracts/service/DelegateManager.sol

```diff
@@ -0,0 +1,293 @@
+      _amount,
+      empty);
+
+    // Update list of delegators to SP if necessary
+    // TODO: Any validation on returned value?
+    updateServiceProviderDelegatorsIfNecessary(delegator, _target);
+
+    // Update amount staked from this delegator to targeted service provider
+    delegateInfo[delegator][_target] += _amount;
+
+    // Update total delegated stake
+    delegatorStakeTotal[delegator] += _amount;
+
+    // Return new total
+    return delegateInfo[delegator][_target];
+  }
+
+  function decreaseDelegatedStake(
+    address _target,
+    uint _amount
+  ) external returns (uint delegateAmount) {
+    address delegator = msg.sender;
+    Staking stakingContract = Staking(
+        registry.getContract(stakingProxyOwnerKey)
+    );
+    bool delegatorRecordExists = updateServiceProviderDelegatorsIfNecessary(delegator, _target);
+    require(delegatorRecordExists, 'Delegator must exist to decrease stake');
+
+    uint currentlyDelegatedToSP = delegateInfo[delegator][_target];
+    require(
+      _amount <= currentlyDelegatedToSP,
+      'Cannot decrease greater than currently staked for this ServiceProvider');
+
+    // Stake on behalf of target service provider 
+    stakingContract.undelegateStakeFor(
+      _target,
+      delegator,
+      _amount,
+      empty);
+
+    // Update amount staked from this delegator to targeted service provider
+    delegateInfo[delegator][_target] -= _amount;
+
+    // Update total delegated stake
+    delegatorStakeTotal[delegator] -= _amount;
+
+    // Remove from delegators list if no delegated stake remaining
+    if (delegateInfo[delegator][_target] == 0) {
+      bool foundDelegator;
+      uint delegatorIndex;
+      for (uint i = 0; i < spDelegates[_target].length; i++) {
+        if (spDelegates[_target][i] == delegator) {
+          foundDelegator = true;
+          delegatorIndex = i;
+        }
+      }
+
+      // Overwrite and shrink delegators list
+      spDelegates[_target][delegatorIndex] = spDelegates[_target][spDelegates[_target].length - 1];
+      spDelegates[_target].length--;
+    }
+
+    // Return new total
+    return delegateInfo[delegator][_target];
+  }
+
+  function makeClaim() external {
+    // address claimer = msg.sender;
+    ServiceProviderFactory spFactory = ServiceProviderFactory(
+        registry.getContract(serviceProviderFactoryKey)
+    );
+
+    // Amount stored in staking contract for owner
+    uint totalBalanceInStaking = Staking(
+        registry.getContract(stakingProxyOwnerKey)
+    ).totalStakedFor(msg.sender);
+    require(totalBalanceInStaking > 0, 'Stake required for claim');
+
+    // Amount in sp factory for claimer
+    uint totalBalanceInSPFactory = spFactory.getServiceProviderStake(msg.sender);
+    require(totalBalanceInSPFactory > 0, 'Service Provider stake required');
+
+    // Amount in delegate manager staked to service provider
+    // TODO: Consider caching this value
+    uint totalBalanceInDelegateManager = 0;
```

Comment:
this is actually totalBalanceInDelegateManagerForSP - Consider caching this value and updating as operations happen. it could save us significant iteration cost in claim and in slash